File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubbleProxyAdapter.java
package closer.vlllage.com.closer.handler.bubble;

import android.content.res.ColorStateList;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.event.EventDetailsHandler;
import closer.vlllage.com.closer.handler.helpers.ImageHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Group;
import jp.wasabeef.picasso.transformations.RoundedCornersTransformation;

public class MapBubbleProxyAdapter extends PoolRecyclerAdapter<MapBubbleProxyAdapter.ProxyMapBubbleViewHolder> {
    private final List<MapBubble> items = new ArrayList<>();
    private MapBubbleView.OnMapBubbleClickListener onClickListener;
    private MapBubble proxyMapBubble;

    public MapBubbleProxyAdapter(PoolMember poolMember, MapBubble proxyMapBubble, MapBubbleView.OnMapBubbleClickListener onClickListener) {
        super(poolMember);
        this.onClickListener = onClickListener;
        this.proxyMapBubble = proxyMapBubble;
    }

    public MapBubbleProxyAdapter setItems(List<MapBubble> items) {
        this.items.clear();
        this.items.addAll(items);
        return this;
    }

    @NonNull
    @Override
    public ProxyMapBubbleViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ProxyMapBubbleViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.map_bubble_proxy_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ProxyMapBubbleViewHolder holder, int position) {
        MapBubble mapBubble = items.get(position);
        switch (mapBubble.getType()) {
            case STATUS:
                holder.click.setBackgroundResource(R.drawable.clickable_blue_4dp);
                holder.photo.setVisibility(View.GONE);
                holder.name.setText(mapBubble.getName() + "\n" + mapBubble.getStatus());
                break;
            case PHYSICAL_GROUP:
                holder.click.setBackgroundResource(R.drawable.clickable_purple_4dp);
                holder.photo.setVisibility(View.VISIBLE);

                if (mapBubble.getTag() != null & mapBubble.getTag() instanceof Group) {
                    Group group = (Group) mapBubble.getTag();
                    if (group.getPhoto() != null) {
                        holder.photo.setColorFilter(null);
                        holder.photo.setImageTintList(ColorStateList.valueOf($(ResourcesHandler.class).getResources().getColor(android.R.color.transparent)));
                        $(ImageHandler.class).get()
                                .load(group.getPhoto() + "?s=32")
                                .fit()
                                .transform(new RoundedCornersTransformation($(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.physicalGroupCorners), 0))
                                .into(holder.photo);
                    } else {
                        holder.photo.setImageResource(R.drawable.ic_wifi_black_24dp);
                        holder.photo.setImageTintList(ColorStateList.valueOf($(ResourcesHandler.class).getResources().getColor(android.R.color.white)));
                    }
                }

                holder.name.setText($(Val.class).of(((Group) mapBubble.getTag()).getName(), $(ResourcesHandler.class).getResources().getString(R.string.app_name)));
                break;
            case EVENT:
                holder.click.setBackgroundResource(R.drawable.clickable_red_4dp);
                holder.photo.setVisibility(View.GONE);

                Event event = ((Event) mapBubble.getTag());

                holder.name.setText(event.getName() + "\n" + $(EventDetailsHandler.class).formatEventDetails(event));

                if (event.isPublic()) {
                    holder.name.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, 0, 0);
                } else {
                    holder.name.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_lock_black_18dp, 0, 0, 0);
                }
                break;
        }

        holder.itemView.setOnClickListener(view -> {
            mapBubble.setView(view);
            onClickListener.onMapBubbleClick(mapBubble);
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class ProxyMapBubbleViewHolder extends RecyclerView.ViewHolder {

        View click;
        ImageView photo;
        TextView name;

        public ProxyMapBubbleViewHolder(View itemView) {
            super(itemView);
            click = itemView.findViewById(R.id.click);
            photo = itemView.findViewById(R.id.photo);
            name = itemView.findViewById(R.id.name);
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/bubble/MapBubbleProxyAdapter.kt
package closer.vlllage.com.closer.handler.bubble

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.extensions.visible
import closer.vlllage.com.closer.handler.event.EventDetailsHandler
import closer.vlllage.com.closer.handler.group.PhysicalGroupHandler
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter
import closer.vlllage.com.closer.store.models.*
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.queatz.on.On
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers

class MapBubbleProxyAdapter(on: On, private val onClickListener: (MapBubble) -> Unit)
    : PoolRecyclerAdapter<MapBubbleProxyAdapter.ProxyMapBubbleViewHolder>(on) {

    var items = mutableListOf<MapBubble>()
        set(value) {
            field.clear()
            field.addAll(value)
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProxyMapBubbleViewHolder {
        return ProxyMapBubbleViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.map_bubble_proxy_item, parent, false))
    }

    override fun onBindViewHolder(holder: ProxyMapBubbleViewHolder, position: Int) {
        val mapBubble = items[position]
        holder.on = On(on).apply {
            use<DisposableHandler>()
        }

        when (mapBubble.type) {
            BubbleType.STATUS -> {
                holder.name.setTextColor(on<ResourcesHandler>().resources.getColor(R.color.textInverse))
                holder.info.setTextColor(on<ResourcesHandler>().resources.getColor(R.color.textHintInverse))
            }
            else -> {
                holder.name.setTextColor(on<ResourcesHandler>().resources.getColor(R.color.text))
                holder.info.setTextColor(on<ResourcesHandler>().resources.getColor(R.color.textHint))
            }
        }

        holder.background.visible = false
        holder.background.clipToOutline = true

        when (mapBubble.type) {
            BubbleType.STATUS -> {
                holder.click.setBackgroundResource(R.drawable.clickable_white_8dp)
                holder.photo.visible = true
                holder.photo.colorFilter = null
                holder.photo.imageTintList = ColorStateList.valueOf(on<ResourcesHandler>().resources.getColor(android.R.color.transparent))
                holder.photo.setImageResource(R.drawable.ic_person_black_24dp)
                holder.name.text = "${mapBubble.name}${on<PhoneDetailsHandler>().detailsOf(mapBubble.tag as Phone)}\n${mapBubble.status ?: ""}"
                holder.info.visible = true
                holder.info.text = on<TimeStr>().pretty((mapBubble.tag as Phone).updated)

                if (mapBubble.tag is Phone) {
                    val phone = mapBubble.tag as Phone
                    if (!phone.photo.isNullOrBlank()) {
                        on<PhotoHelper>().loadCircle(holder.photo, phone.photo!!)
                    }
                }
            }
            BubbleType.SUGGESTION -> {
                holder.click.setBackgroundResource(R.drawable.clickable_blue_light_8dp)
                holder.photo.visible = true
                holder.photo.colorFilter = null
                holder.photo.imageTintList = ColorStateList.valueOf(on<ResourcesHandler>().resources.getColor(android.R.color.white))
                holder.photo.setImageResource(R.drawable.ic_edit_location_black_24dp)
                holder.name.text = mapBubble.status
                holder.info.visible = true
                holder.info.text = on<TimeStr>().prettyDate((mapBubble.tag as Suggestion).created)
            }
            BubbleType.PHYSICAL_GROUP -> {
                holder.click.setBackgroundResource(R.drawable.clickable_purple_8dp)
                holder.photo.visible = true
                holder.info.visible = false

                if ((mapBubble.tag != null) and (mapBubble.tag is Group)) {
                    val group = mapBubble.tag as Group

                    if (group.photo != null) {
                        holder.background.setImageDrawable(null)
                        holder.background.visible = true
                        on<ImageHandler>().get()
                                .load(group.photo!! + "?s=128")
                                .transition(DrawableTransitionOptions.withCrossFade())
                                .into(holder.background)
                    }

                    holder.photo.setImageResource(R.drawable.ic_chat_black_24dp)
                    holder.photo.imageTintList = ColorStateList.valueOf(on<ResourcesHandler>().resources.getColor(android.R.color.white))
                }

                on<PhysicalGroupHandler>().physicalGroupName((mapBubble.tag as Group))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                    holder.name.text = it
                }, {}).also {
                    holder.on<DisposableHandler>().add(it)
                }
            }
            BubbleType.EVENT -> {
                holder.click.setBackgroundResource(R.drawable.clickable_red_8dp)
                holder.photo.visible = false
                holder.info.visible = false

                val event = mapBubble.tag as Event

                holder.name.text = "${event.name}\n${on<EventDetailsHandler>().formatEventDetails(event)}"

                if (event.isPublic) {
                    holder.name.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, 0, 0)
                } else {
                    holder.name.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_group_black_18dp, 0, 0, 0)
                }
            }
            else -> {
                holder.name.visible = true
                holder.name.text = on<ResourcesHandler>().resources.getString(R.string.unknown)
            }
        }

        holder.itemView.setOnClickListener { view ->
            mapBubble.view = view
            onClickListener.invoke(mapBubble)
        }
    }

    override fun onViewRecycled(holder: ProxyMapBubbleViewHolder) {
        holder.on.off()
    }

    override fun getItemCount() = items.size

    inner class ProxyMapBubbleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var click: View = itemView.findViewById(R.id.click)
        var photo: ImageView = itemView.findViewById(R.id.photo)
        var background: ImageView = itemView.findViewById(R.id.background)
        var name: TextView = itemView.findViewById(R.id.name)
        var info: TextView = itemView.findViewById(R.id.info)
        lateinit var on: On
    }
}
